### PR TITLE
Glossary [work in progress]

### PR DESCRIPTION
**Pull request by Sebastian-Lemling**
**working on https://github.com/privacyidea/privacyidea/issues/2783:**
* [x] application plugins bcf7469b60205d397baa6cd4e642cdf4b307cf5a
* [x] administrator 393aaa53b876c0b022a49e7cc060657019a0d419
* [x] opt pin 59c7e3ba0d846ea138f49a64b7cdf0d2a1ae26bc
* [x] otp value 59c7e3ba0d846ea138f49a64b7cdf0d2a1ae26bc
* [x] time step 83b4e652925dcd0a265e6d293b868a10446ea485
* [x] time window f3b307fc98ec7f6c553badc2db2e81efe83e1349
* [x] userinfo 4604f056c95f08f7d4081d91d11da81028e8a456
* [x] userstore 2c35e3b21aa2b9117241dd9dd54a660c53779851
* [x] disabled token bbfb8480476a6676d8744c610ff7497a253b870a
* [x] revoked token 55e4663f2922a01b0029c4c753ebb9cc36f29219
* [x] orphaned token 6eccd6a3f600ba3b34434967ecf930bdcd55a0e6
* [x] resolver 3dd0b0e9d02091e43f83b07f39d639c39186d625

## Pool

* [x] token cee46fcb85f8c6ea0ff2d2fb1d1a05489001d8e2
* [x] tokeninfo c2bd1ef60ebef7db18932497e3f67d170a61f44f
* [x] audit cfffbe99ef68f5cb54e6a0d70e2f4ff8e14ff9d2
* [x] token owner 9fc2e308cb944e485732bebb7d20d0792624fd49
* [x] MaxFail and FailCount 69f0ebe71b563c8f204594056ad9228d1b8759a5
* [x] userid 090fa576008fd663f88b0d9a760bed9d20b01000
* [x] split @ sign f1380280434cc5cc2a0e4398bc873469d7e7e6a7
* [x] subscription 77289acedf1f72f252db59d9041f3a21caeea8ab
* [x] rollout state 0764796bed62b16c729c86ae758effe3fd010977
* [x] seed 59c7e3ba0d846ea138f49a64b7cdf0d2a1ae26bc
* [x] challenge f427559b4915bb105da23a0b1f28d2e7f27e16eb
* [x] multi challenge f427559b4915bb105da23a0b1f28d2e7f27e16eb
* [x] scope 6989f65163dde683ac9e9e9bd91670df185a565c
* [x] realm d5343bde64ffb62feecfd52c223cc17eb7350378
* [x] event 43c8ea7afe26210ea958b5ea95d35b10f0cf81e8
* [x] WebUI 406158f3db456393660ae6be321e56ea2351e61e
* [x] extended attribute mapping -> RADIUS d8d01d47bee121130c2b501945602af979d99ef2
* [x] policies conditions  aaa7a6b4d86f49804fb998bc7fd22beee45858b2
* [x] custom user attributes 6e0e41e1dbecbc867bb6a1b53270323252ab7bfb
* [x] count/counter 59c7e3ba0d846ea138f49a64b7cdf0d2a1ae26bc

### CODE REVIEW

File: doc/glossary/index.rst [hunk 1]
```diff
@@ -31,3 +31,25 @@ Glossary
...
+        * :ref:`usersview`
+        * :ref:`machines`
+        * :ref:`config`
+        * :ref:`audit`
+        * :ref:`components`
+
+   Time Step
+        A TOTP token can have a timestep of 30/60 seconds. It can still be used, if this 30/60 seconds are over.
+        Technically the timestep is the divider by which the seconds since 1.1.1970 (unix system time)
+        are devided to calculate the OTP value.
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 7e5080ba02c542f964a25fef2c19723305f74a08.

File: doc/glossary/index.rst [hunk 2]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        * :ref:`config`
+        * :ref:`audit`
+        * :ref:`components`
+
+   Time Step
+        A TOTP token can have a timestep of 30/60 seconds. It can still be used, if this 30/60 seconds are over.
+        Technically the timestep is the divider by which the seconds since 1.1.1970 (unix system time)
+        are divided to calculate the OTP value.
+
+        The ``timestep`` is different to the :ref:`Time Window`.
```
Comment (cornelinux) [inline doc/glossary/index.rst:55]: I think you do not have defined a reference anchor for "Time Window".

File: doc/glossary/index.rst [hunk 3]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        not be modified anymore. It can only be deleted. Certain token types like certificate may
+        define special actions when revoking a token.
+
+   Orphaned Token
+        An orphaned token means, that it has a user assigned, but the user does not exist in the
+        user store (anymore).
+
+   Disabled Token
+        Tokens can be ``disabled``. Disabled tokens still belong to the assigned user but those tokens
+        can not be used to authenticate. ``Disabled tokens can be enabled again``.
```
Comment (cornelinux) [inline doc/glossary/index.rst:72]: Just technically: The formatting makes this text a "command" format. You prabably want to use:

~~~~
*Disabled tokens can be enabled again*.
~~~~

File: doc/glossary/index.rst [hunk 4]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+
+   Tokeninfo
+        The table “tokeninfo” is used to store additional, long information that is
+        specific to the :ref:`tokentypes`. E.g. the tokentype “TOTP” has additional entries
+        in the tokeninfo table for “timeStep” and “timeWindow”, which are stored in the
+        column “Key” and “Value”.
+
+        The ``tokeninfo`` is reference by the foreign key to the “token” table.
+
+        Token info can be viewed and partially edited in the GUI. In addition, the
```
Comment (cornelinux) [inline doc/glossary/index.rst:118]: GUI -> WebUI (consistent naming)

File: doc/glossary/index.rst [hunk 5]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        Token-Janitor can be used to output token info, filter for tokens that have
+        specific tokeninfo and set user-defined tokeninfos.
+
+   Token
+        All token information is stored in an SQL database, while you may choose, which database
+        you want to use. privacyIDEA uses SQLAlchemy to map the database to internal objects.
+        Thus you may choose to run privacyIDEA with SQLite, MySQL, PostgreSQL, Oracle,
+        DB2 or other database.
```
Comment (cornelinux) [inline doc/glossary/index.rst:126]: I think all this database stuff does not belong to explaining what a token is.

File: doc/glossary/index.rst [hunk 6]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        Tools which facilitate and automate token enrollment are found in :ref:`enrollment_tools`.
+
+        * :ref:`authentication_modes`
+        * :ref:`supported_tokens`
+        * :ref:`tokentypes`
+
+   Audit
+        The systems provides a sophisticated audit log, that can be viewed in the WebUI.
+
+        The Audit log lists all events the server registers.
```
Comment (cornelinux) [inline doc/glossary/index.rst:142]: registers?

File: doc/glossary/index.rst [hunk 7]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+
+   Tokenowner
+        The owner of a token is the user for whom the token was rolled out.
+
+   FailCount
+   MaxFail
+        The FailCount count the number of failed login attempts.
+
+        If the login fail counter reaches the ``MaxFail`` the user can not login
+        with this token anymore.
```
Comment (cornelinux) [inline doc/glossary/index.rst:163]: note, that the fail counter is not increased anymore, if it has reached the maxfail.

File: doc/glossary/index.rst [hunk 8]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+
+        How a user is related to a realm is described here: :ref:`relate_realm`
+
+        This option also affects the login via the :ref:`rest_auth`
+
+   Rollout State
+        A token can be rolled out in several steps like the 2step HOTP/TOTP token.
+        In this case the attribute ``“rollout_state”`` of the token contains certain values
+        like ``‘clientwait’ or ‘enrolled’``. This way actions can be triggered, depending
+        on the step during an enrollment process.
```
Comment (cornelinux) [inline doc/glossary/index.rst:186]: Rollout States are:     

'clientwait'
 # The rollout is pending in the backend, like CSRs that need to be approved
'pending'
 # This means the user needs to authenticate to verify that the token was successfully enrolled.
'verify'
'enrolled'
'failed'

Note, not all tokens have the rollout state "enrolled" set consistently. An empty rollout state means "enrolled".

File: doc/glossary/index.rst [hunk 9]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+
+   Custom User Attributes
+        The table ``“customuserattribute”`` is used to store additional, custom attributes for users.
+
+        A user is identified by the user_id, the resolver_id and the realm_id.
+        The additional attributes are stored in Key and Value.
+        The Type can hold extra information like e.g. an encrypted value / password.
```
Comment (cornelinux) [inline doc/glossary/index.rst:193]: The important part is, that the privacyIDEA administrator, who is often not allowed to manage active directory, can **add** attributes to the users. 
Normally user attributes are only read from the user store.

https://www.youtube.com/watch?v=QU30r99oE3o

File: doc/glossary/index.rst [hunk 10]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+
+        * USER = 'user'
+
+        * WEBUI = 'webui'
+
+        scope takes only one value.
+
+   Realms
+        Users need to be in realms to have tokens assigned. A user, who is not
+        member of a realm can not have a token assigned and can not authenticate.
```
Comment (cornelinux) [inline doc/glossary/index.rst:223]: realms are ment for general logical user grouping.

File: doc/glossary/index.rst [hunk 11]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        .. figure:: /eventhandler/event-list.png
+           :width: 500
+
+           *An action is bound to the event* token_init.
+
+   Radius Attribute Mapping
+        The Radius plugin can use information from the ``detail`` section
+        (see :ref:`rest_validate`) of the
+        privacyIDEA response to map these values to arbitrary RADIUS Attribute-Value
+        pairs.
```
Comment (cornelinux) [inline doc/glossary/index.rst:259]: https://www.youtube.com/watch?v=uERhuCLxz0o

File: doc/glossary/index.rst [hunk 12]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        Web services often use SQL databases as user store.
+
+   Userinfo
+        These are the user attributes as they are determined by the respective resolver.
+        This is configured via the attribute mappings of resolvers (see :ref:`useridresolvers`).
+
+   Challenge
+   Multi Challenge
+        If a user wants to authenticate with his username and password, privacyIDEA will check whether an active
+        challenge response token exists for this user. In this case, the challenge is triggered and privacyIDEA expects a response.
```
Comment (cornelinux) [inline doc/glossary/index.rst:359]: Well, often people do not know how things (like SMS) are triggered.

It is important to point out, that challenges are triggered by 

* the user entering the PIN/Password of the token
* or programatically via a call to /validate/triggerchallenge

File: doc/glossary/index.rst [hunk 13]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        If the user now gives the answer expected from the server, the response is accepted and the authentication is successful.
+
+        Multi Challenge is basically a chain of challenges. If, for example, a PIN is stored for a challenge response token, after
+        the first successful challenge, a second one is triggered, in which the PIN is expected as an answer.
+        If the PIN matches, the second challenge is also successful and the user is successfully authenticated.
```
Comment (cornelinux) [inline doc/glossary/index.rst:364]: Unfortunately this is wong! The PIN triggers the first challenge.

Multichallenge can be used to reset a PIN or with the four-eyes-token.

File: doc/glossary/index.rst [hunk 14]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+
+        Conditions can be added to a policy via the WebUI. In order for a policy to
+        take effect during the processing of a request, the request has to match not
+        only the ordinary policy attributes (see :ref:`policies`), but also *all*
+        additionally defined conditions that are currently active. If no active
+        conditions are defined, only the ordinary policy attributes are taken into
+        account.
+
+   Application Plugins
+        privacyIDEA comes with application plugins. These are plugins for
```
Comment (cornelinux) [inline doc/glossary/index.rst:379]: I would say: Plugins are available.

"comes with" could cause false expectations.

File: doc/glossary/index.rst [hunk 15]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        privacyIDEA is enterprise software. Managing lots of authentication devices for lots of users is a task that occurs in a company network.
+        privacyIDEA is licensed under an Open Source license. This guarantees, that a company using the Open Source software privacyIDEA can use this
+        software for life. In contrast prorpietary software or software-as-a-service (SaaS) can be changed, billed differently or even completely deleted.
+        You could not do anything about it. The Open Source privacyIDEA is under your control – forever.
+        The Open Source license dos not mean that a company has no costs in regards to two factor authentication.
+        At least they need to pay the administrator.
+        In any case the Open Source license states that this software comes without any warranty. A company using privacyIDEA needs to be aware of this.
+
+        For the product privacyIDEA we provide the suitable `support <https://netknights.it/en/produkte/privacyidea/>`_ with a defined response time and with fixed costs.
+
```
Comment (cornelinux) [inline doc/glossary/index.rst:400]: The paragraph is missing the last information, what the "subscription" is. A subscription for support and enterprise packaging.

File: doc/glossary/index.rst (hunk 4 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in f61a46945fb583cac2e66ad33600bb26eae87ded.

File: doc/glossary/index.rst (hunk 2 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in deb552c3c5b4a7656793fc1d800c15c2286195d4.

File: doc/glossary/index.rst (hunk 3 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 0dd7436e2018a6334c568484855c62c2f2a2e0da.

File: doc/glossary/index.rst [hunk 16]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+        * :ref:`flatfile_resolver`
+        * :ref:`ldap_resolver`
+        * :ref:`sql_resolver`
+        * :ref:`scim_resolver`
+        * :ref:`http_resolver`
+
+   Resolver(Machine)
+        Machine Resolvers are used to find machines in directories like LDAP,
+        Active Directory, puppet, salt, or the /etc/hosts file.
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 31c3603ab6212d9ea232223a9a5e92bf3bc716b3.

File: doc/glossary/index.rst (hunk 5 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in db6f287c3112c04a59b67b99766c2f3fe74a2385.

File: doc/glossary/index.rst (hunk 6 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 77a9cf862391a2715de0ba765c304263757524b9.

File: doc/glossary/index.rst (hunk 7 above)
Comment (NNLemling) [reply doc/glossary/index.rst:159]: done in e3573cd4cba201789de8327cd542213de2f3c657.

File: doc/glossary/index.rst (hunk 8 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 88b3cb4811dc144e9bdec74d2e232fa843076b2d.

File: doc/glossary/index.rst (hunk 9 above)
Comment (NNLemling) [reply doc/glossary/index.rst:222]: done in 4c9671867f1721aeb42e3525f74855c4335a9440.

File: doc/glossary/index.rst (hunk 10 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 3654d0695eb1bf3fee0a79024198bf292d2d29eb.

File: doc/glossary/index.rst (hunk 11 above)
Comment (NNLemling) [reply doc/glossary/index.rst:288]: done in f4e110ab89c11f9bc7603cc003a4c8d438ac1665.

File: doc/glossary/index.rst (hunk 12 above)
Comment (NNLemling) [reply doc/glossary/index.rst:390]: done in f9f28db97f886ae132df666cc409008769c9b311.

File: doc/glossary/index.rst (hunk 13 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in f9f28db97f886ae132df666cc409008769c9b311.

File: doc/glossary/index.rst [hunk 17]
```diff
@@ -31,3 +31,374 @@ Glossary
...
+   Multi Challenge
+        If a user wants to authenticate with his username and password, privacyIDEA will check whether an active
+        challenge response token exists for this user. In this case, the challenge is triggered and privacyIDEA expects a response.
+        If the user now gives the answer expected from the server, the response is accepted and the authentication is successful.
+
+        Multi Challenge is basically a chain of challenges. If, for example, a PIN is stored for a challenge response token, after
+        the first successful challenge, a second one is triggered, in which the PIN is expected as an answer.
+        If the PIN matches, the second challenge is also successful and the user is successfully authenticated.
+
+   Policy Conditions
```
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 62d5caa739215a3812239e418d0733ebb6f7dcbf.

File: doc/glossary/index.rst (hunk 14 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 53bb67399db07edeb0402718201637498de5b8a2.

File: doc/glossary/index.rst (hunk 15 above)
Comment (NNLemling) [reply doc/glossary/index.rst:440]: done in 6a521e1c8da886ba9e6f80c2c00ba8eff01eab42.

File: doc/glossary/index.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/glossary/index.rst]: Well, but privacyIDEA can not read machines defined in puppet. So you need to remove puppet and salt!

File: doc/glossary/index.rst (hunk 8 above)
Comment (cornelinux) [reply doc/glossary/index.rst]: See and fix it.

File: doc/glossary/index.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 4cd7f887f6b180edb6e706ab844513c5ae2f4ccf.

File: doc/glossary/index.rst (hunk 8 above)
Comment (NNLemling) [reply doc/glossary/index.rst]: done in 9e1730c84ea3c457b4b8ed7ec1956ef9876a77ed.

**Review by cornelinux — APPROVED** [review]

**Pull request merged**